File: NativeAPI/Native-API/SUT/csharp/DataBusClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Metamoto.Services;
using Metamoto.Messages;
using Grpc.Core;

class DataBusClient {
  private const string VEHICLE_STATE_TOPIC = "VehicleState";
  private const string VEHICLE_CONTROLS_TOPIC = "VehicleControls";
  
  private Channel _channel;
  private DataBus.DataBusClient _client;
  private string _name;
  private string _egoVehicleId;
  
  public bool Initialize(string name, string uri, string egoVehicleId) {
    try {
      _channel = new Channel(uri, ChannelCredentials.Insecure);
      _client = new DataBus.DataBusClient(_channel);    
    } catch (RpcException rpcEx) {
      Console.WriteLine("DataBusClient::Initialize: " + rpcEx.Status.StatusCode + ", " + rpcEx.Status.Detail);
      return false;
    } catch (System.Exception ex) {
      Console.WriteLine("DataBusClient::Initialize: " + ex.Message);
      return false;
    }    
    
    _name = name;
    _egoVehicleId = egoVehicleId;
    
    return true;
  }  

  public void Close() {
    if ((_client == null) || (_channel == null)) {
      return;
    }
    
    _channel.ShutdownAsync().Wait();
    _channel = null;
    _client = null;    
  }  
  
  public bool Subscribe(string topic) {
    if ((_client == null) || (_channel == null)) {
      return false;
    }

    DataBusSubscribeRequest request = new DataBusSubscribeRequest();
    request.Topic = topic;
    request.ReaderName = _name;
    request.EgoVehicleId = _egoVehicleId;
    
    try {
      DataBusSubscribeReply reply = _client.Subscribe(request);
      if (reply != null) {
        Console.WriteLine("DataBusClient::Subscribe: Subscribed to " + request.Topic);
        return true;
      } else {
        Console.WriteLine("DataBusClient::Subscribe: Data bus subscribe reply is null.");
      }
    } catch (RpcException rpcEx) {
      Console.WriteLine("DataBusClient::Subscribe: RPC Exception: " + rpcEx.Status.StatusCode + ", " + rpcEx.Status.Detail);
    } catch (System.Exception ex) {
      Console.WriteLine("DataBusClient::Subscribe: System Exception: " + ex.Message);
    }
    
    return false;
  }    

  public bool Publish(string topic) {
    if ((_client == null) || (_channel == null)) {
      return false;
    }

    DataBusPublishRequest request = new DataBusPublishRequest();
    request.Topic = topic;
    request.WriterName = _name;
    request.EgoVehicleId = _egoVehicleId;
    
    try {
      DataBusPublishReply reply = _client.Publish(request);
      if (reply != null) {
        Console.WriteLine("DataBusClient::Publish: Publishing " + request.Topic);
        return true;
      } else {
        Console.WriteLine("DataBusClient::Publish: Data bus publish reply is null.");
      }
    } catch (RpcException rpcEx) {
      Console.WriteLine("DataBusClient::Publish: RPC Exception: " + rpcEx.Status.StatusCode + ", " + rpcEx.Status.Detail);
    } catch (System.Exception ex) {
      Console.WriteLine("DataBusClient::Publish: System Exception: " + ex.Message);
    }
    
    return false;
  }    
  
  public DataBusMessage Read(double t, string topic) {
    if ((_client == null) || (_channel == null)) {
      return null;
    }
    
    DataBusReadRequest request = new DataBusReadRequest();
    request.Topic = topic;
    request.ReaderName = _name;
    request.EgoVehicleId = _egoVehicleId;
    request.Time = t;
    
    DataBusReadReply reply = null;
    
    try {
      reply = _client.Read(request);
    } catch (RpcException rpcEx) {
      Console.WriteLine("DataBusClient::ReadVehicleState: RPC Exception: " + rpcEx.Status.StatusCode + ", " + rpcEx.Status.Detail);
      return null;
    } catch (System.Exception ex) {
      Console.WriteLine("DataBusClient::ReadVehicleState: System Exception: " + ex.Message);
      return null;
    }
    
    return reply.Message;
  }  
  
  public bool Write(DataBusMessage message, string topic) {
    if ((_client == null) || (_channel == null)) {
      return false;
    }
    
    DataBusWriteRequest request = new DataBusWriteRequest();
    request.Topic = topic;
    request.WriterName = _name;
    request.EgoVehicleId = _egoVehicleId;
    request.Message = message;
    
    DataBusWriteReply reply = null;
    
    try {
      reply = _client.Write(request);
    } catch (RpcException rpcEx) {
      Console.WriteLine("WriteVehicleControls: RPC Exception: " + rpcEx.Status.StatusCode + ", " + rpcEx.Status.Detail);
      return false;
    } catch (System.Exception ex) {
      Console.WriteLine("WriteVehicleControls: System Exception: " + ex.Message);
      return false;
    }    

    return true;
  }
}

File: NativeAPI/Native-API/SUT/cpp/data_bus_client.cc
#include <exception>
#include "data_bus_client.h"

using namespace std;
using namespace metamoto::messages;
using namespace metamoto::services;
using metamoto::types::NormalDistribution;


bool DataBusClient::Publish(const string& topic, const string& writerName) {
  DataBusPublishRequest request;
  DataBusPublishReply reply;
  
  grpc::ClientContext context;
  request.mutable_latency()->set_mean(0.1);
  request.mutable_latency()->set_std_dev(0.04);
  request.set_topic(topic);
  request.set_writer_name(writerName);
  request.set_ego_vehicle_id(_egoVehicleId);
  grpc::Status status;

  try {
    status = _stub->Publish(&context, request, &reply);
  } catch (exception& e) {
    cout << e.what() << endl;
    return false;
  }
  
  if (status.ok()) {
    cout << "DataBus Publish to topic " << topic << "." << endl;
    return true;
  } else {
    cout << "DataBus Publish failed to topic " << topic << ": " << status.error_message() << endl;
    return false;
  }
}

bool DataBusClient::Subscribe(const string& topic, const string& readerName) {
  DataBusSubscribeRequest request;
  DataBusSubscribeReply reply;

  grpc::ClientContext context;
  request.set_topic(topic);
  request.set_reader_name(readerName);
  request.set_ego_vehicle_id(_egoVehicleId);
  
  grpc::Status status = _stub->Subscribe(&context, request, &reply);

  if (status.ok()) {
    cout << "Subscribed to dataBus topic " << topic << "." << endl;
    return true;
  } else {
    cout << "Failed to Subscribe to dataBus topic " << topic << ": " << status.error_message() << endl;
    return false;
  }
}

bool DataBusClient::Write(const string& topic, const string& writerName, const DataBusMessage& message) {
  DataBusWriteRequest request;
  DataBusWriteReply reply;

  grpc::ClientContext context;
  request.set_topic(topic);
  request.set_writer_name(writerName);
  request.set_ego_vehicle_id(_egoVehicleId);
  request.mutable_message()->CopyFrom(message);
  
  grpc::Status status = _stub->Write(&context, request, &reply);

  if (status.ok()) {
    cout << "DataBus write for topic " << topic << "." << endl;
    return true;
  } else {
    cout << "DataBus Write failed for topic " << topic << ": " << status.error_message() << endl;
    return false;
  }
}

unique_ptr<DataBusMessage> DataBusClient::Read(const string& topic, 
                                               const string& readerName, 
                                               double time) {
  DataBusReadRequest request;
  DataBusReadReply reply;

  grpc::ClientContext context;
  request.set_topic(topic);
  request.set_time(time);
  request.set_reader_name(readerName);
  request.set_ego_vehicle_id(_egoVehicleId);

  grpc::Status status = _stub->Read(&context, request, &reply);

  if (status.ok()) {
    cout << "DataBus Read for topic " << topic << "." << endl;
    return unique_ptr<DataBusMessage>(new DataBusMessage(reply.message()));
  } else {
    cout << "DataBus Read failed for topic " << topic << ": " << status.error_message() << endl;
    return nullptr;
  }
}

unique_ptr<vector<DataBusMessage>> DataBusClient::ReadAll(const string& topic, 
                                                          const string& readerName, 
                                                          double time) {
  DataBusReadAllRequest request;
  DataBusReadAllReply reply;

  grpc::ClientContext context;
  request.set_topic(topic);
  request.set_time(time);
  request.set_reader_name(readerName);
  request.set_ego_vehicle_id(_egoVehicleId);

  grpc::Status status = _stub->ReadAll(&context, request, &reply);

  if (status.ok()) {
    cout << "DataBus ReadAll." << endl;
    return unique_ptr<vector<DataBusMessage>>(new vector<DataBusMessage>(reply.messages().begin(),
                                                                         reply.messages().end()));
  } else {
    cout << "DataBus ReadAll failed: " << status.error_message() << endl;
    return nullptr;
  }
}

unique_ptr<DataBusMessage> DataBusClient::ReadLast(const string& topic, 
                                                   const string& readerName, 
                                                   double time) {
  DataBusReadLastRequest request;
  DataBusReadLastReply reply;

  grpc::ClientContext context;
  request.set_topic(topic);
  request.set_time(time);
  request.set_reader_name(readerName);
  request.set_ego_vehicle_id(_egoVehicleId);

  grpc::Status status = _stub->ReadLast(&context, request, &reply);

  if (status.ok()) {
    cout << "DataBus ReadLast for topic " << topic << "." << endl;
    return unique_ptr<DataBusMessage>(new DataBusMessage(reply.message()));
  } else {
    cout << "DataBus ReadLast failed for topic " << topic << ": " << status.error_message() << endl;
    return nullptr;
  }
}
